File: 05_Generalized_Linear_Models.R
## load packages
library(tidyverse)
library(nflfastR)
library(broom)

## load data and filter data
chap_5_file <- "./data/pbp_r_chap_5.csv"
if (!file.exists(chap_5_file)) {
    pbp_r <- load_pbp(2016:2022)
    write_csv(pbp_r, chap_5_file)
} else {
    pbp_r <- read_csv(chap_5_file)
}


pbp_r_pass <-
    pbp_r |>
    filter(
        play_type == "pass" & !is.na(passer_id) & !is.na(air_yards)
    )

## filter more and then plot
pass_pct_r <-
    pbp_r_pass |>
    filter(0 < air_yards & air_yards <= 20) |>
    group_by(air_yards) |>
    summarize(comp_pct = mean(complete_pass), .groups = "drop")

pass_pct_r |>
    ggplot(aes(x = air_yards, y = comp_pct)) +
    geom_point() +
    stat_smooth(method = "lm") +
    theme_bw() +
    ylab("Percent completion") +
    xlab("Air yards")

## building a glm
complete_ay_r <-
    glm(complete_pass ~ air_yards,
        data = pbp_r_pass, family = "binomial"
    )
summary(complete_ay_r)

## logistic plot
ggplot(data = pbp_r_pass, aes(x = air_yards, y = complete_pass)) +
    geom_jitter(height = 0.05, width = 0, alpha = 0.05) +
    stat_smooth(method = "glm", method.args = list(family = "binomial")) +
    theme_bw() +
    ylab("Completed pass (1 = yes, 0 = no)") +
    xlab("air yards")

## glm completion percentage predicted
pbp_r_pass <-
    pbp_r_pass |>
    mutate(
        exp_completion = predict(complete_ay_r, type = "resp"),
        cpoe = complete_pass - exp_completion
    )

## cpoe
pbp_r_pass |>
    group_by(season, passer_id, passer) |>
    summarize(
        n = n(),
        cpoe = mean(cpoe, na.rm = TRUE),
        compl = mean(complete_pass, na.rm = TRUE),
        .groups = "drop"
    ) |>
    filter(n >= 100) |>
    arrange(-cpoe) |>
    print(n = 20)

## remove missing data and format
pbp_r_pass_no_miss <-
    pbp_r_pass |>
    mutate(
        down = factor(down),
        qb_hit = factor(qb_hit)
    ) |>
    filter(
        complete.cases(
            down, qb_hit, complete_pass, ydstogo,
            yardline_100, air_yards, pass_location, qb_hit
        )
    )

## run model and save outputs
complete_more_r <-
    pbp_r_pass_no_miss |>
    glm(
        formula = complete_pass ~ down * ydstogo +
            yardline_100 + air_yards + pass_location + qb_hit,
        family = "binomial"
    )

## cpoe
pbp_r_pass_no_miss <-
    pbp_r_pass_no_miss |>
    mutate(
        exp_completion = predict(complete_more_r, type = "resp"),
        cpoe = complete_pass - exp_completion
    )

## summarize data
cpoe_more_r <-
    pbp_r_pass_no_miss |>
    group_by(season, passer_id, passer) |>
    summarize(
        n = n(),
        cpoe = mean(cpoe, na.rm = TRUE),
        compl = mean(complete_pass),
        exp_completion = mean(exp_completion),
        .groups = "drop"
    ) |>
    filter(n > 100)

## print outputs
cpoe_more_r |>
    arrange(-cpoe) |>
    print(n = 20)

## stability
# create current dataframe
cpoe_now_r <-
    cpoe_more_r |>
    select(-n)

# create last-year's dataframe
# and add 1 to season
cpoe_last_r <-
    cpoe_more_r |>
    select(-n) |>
    mutate(season = season + 1) |>
    rename(
        cpoe_last = cpoe, compl_last = compl,
        exp_completion_last = exp_completion
    )

# merge together
cpoe_lag_r <-
    cpoe_now_r |>
    inner_join(cpoe_last_r,
        by = c("passer_id", "passer", "season")
    ) |>
    ungroup()

## pass completion stability
cpoe_lag_r |>
    select(compl_last, compl) |>
    cor(use = "complete.obs")

## CPOE stability
cpoe_lag_r |>
    select(cpoe_last, cpoe) |>
    cor(use = "complete.obs")

## stability expected completions
cpoe_lag_r |>
    select(exp_completion_last, exp_completion) |>
    cor(use = "complete.obs")

## odds ratios
complete_ay_r |>
    tidy(exponentiate = TRUE, conf.int = TRUE)

## summarize raw data
pbp_r_pass |>
    summarize(comp_pct = mean(complete_pass)) |>
    mutate(odds = comp_pct / (1 - comp_pct), log_odds = log(odds))

## global intercept
complete_global_r <-
    glm(complete_pass ~ 1,
        data = pbp_r_pass,
        family = "binomial"
    )

complete_global_r |>
    tidy()

complete_global_r |>
    tidy(exponentiate = TRUE)

File: 05_Generalized_Linear_Models.py
## load packages
import pandas as pd
import numpy as np
import nfl_data_py as nfl
import statsmodels.formula.api as smf
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
import os

## load data and filter data
chap_5_file = "./data/pbp_py_chap_5.csv"
if os.path.isfile(chap_5_file):
    pbp_py = pd.read_csv(chap_5_file, low_memory=False)
else:
    seasons = range(2016, 2022 + 1)
    pbp_py = nfl.import_pbp_data(seasons)
    pbp_py.to_csv(chap_5_file)

pbp_py_pass = pbp_py.query(
    'play_type == "pass" & passer_id.notnull() &' + "air_yards.notnull()"
).reset_index()

# Change theme for chapter
sns.set_theme(style="whitegrid", palette="colorblind")

# Filter more and then plot
pass_pct_py = (
    pbp_py_pass.query("0 < air_yards <= 20")
    .groupby("air_yards")
    .agg({"complete_pass": ["mean"]})
)

pass_pct_py.columns = list(map("_".join, pass_pct_py.columns))

pass_pct_py.reset_index(inplace=True)
pass_pct_py.rename(columns={"complete_pass_mean": "comp_pct"}, inplace=True)

sns.regplot(data=pass_pct_py, x="air_yards", y="comp_pct", line_kws={"color": "red"})
plt.show()

## building a glm
complete_ay_py = smf.glm(
    formula="complete_pass ~ air_yards", data=pbp_py_pass, family=sm.families.Binomial()
).fit()

complete_ay_py.summary()

## logistic plot
sns.regplot(
    data=pbp_py_pass,
    x="air_yards",
    y="complete_pass",
    logistic=True,
    line_kws={"color": "red"},
    scatter_kws={"alpha": 0.05},
)
plt.show()

## glm completion percentage predicted
pbp_py_pass["exp_completion"] = complete_ay_py.predict()
pbp_py_pass["cpoe"] = pbp_py_pass["complete_pass"] - pbp_py_pass["exp_completion"]

## cpoe
cpoe_py = pbp_py_pass.groupby(["season", "passer_id", "passer"]).agg(
    {"cpoe": ["count", "mean"], "complete_pass": ["mean"]}
)
cpoe_py.columns = list(map("_".join, cpoe_py.columns))
cpoe_py.reset_index(inplace=True)
cpoe_py = cpoe_py.rename(
    columns={"cpoe_count": "n", "cpoe_mean": "cpoe", "complete_pass_mean": "compl"}
).query("n > 100")
print(cpoe_py.sort_values("cpoe", ascending=False))

## remove missing data and format data
pbp_py_pass["down"] = pbp_py_pass["down"].astype(str)
pbp_py_pass["qb_hit"] = pbp_py_pass["qb_hit"].astype(str)
pbp_py_pass_no_miss = pbp_py_pass[
    [
        "passer",
        "passer_id",
        "season",
        "down",
        "qb_hit",
        "complete_pass",
        "ydstogo",
        "yardline_100",
        "air_yards",
        "pass_location",
    ]
].dropna(axis=0)

## build and fit model
complete_more_py = smf.glm(
    formula="complete_pass ~ down * ydstogo + "
    + "yardline_100 + air_yards + "
    + "pass_location + qb_hit",
    data=pbp_py_pass_no_miss,
    family=sm.families.Binomial(),
).fit()

## extract output and calculate CPOE
pbp_py_pass_no_miss["exp_completion"] = complete_more_py.predict()
pbp_py_pass_no_miss["cpoe"] = (
    pbp_py_pass_no_miss["complete_pass"] - pbp_py_pass_no_miss["exp_completion"]
)

## summarize outputs, reformat, and rename
cpoe_py_more = pbp_py_pass_no_miss.groupby(["season", "passer_id", "passer"]).agg(
    {"cpoe": ["count", "mean"], "complete_pass": ["mean"], "exp_completion": ["mean"]}
)

cpoe_py_more.columns = list(map("_".join, cpoe_py_more.columns))
cpoe_py_more.reset_index(inplace=True)

cpoe_py_more = cpoe_py_more.rename(
    columns={
        "cpoe_count": "n",
        "cpoe_mean": "cpoe",
        "complete_pass_mean": "compl",
        "exp_completion_mean": "exp_completion",
    }
).query("n > 100")

## print outputs
print(cpoe_py_more.sort_values("cpoe", ascending=False))

## stability
#  keep only the columns needed
cols_keep = ["season", "passer_id", "passer", "cpoe", "compl", "exp_completion"]

# create current dataframe
cpoe_now_py = cpoe_py_more[cols_keep].copy()

# create last-year's dataframe
cpoe_last_py = cpoe_now_py[cols_keep].copy()

# rename columns
cpoe_last_py.rename(
    columns={
        "cpoe": "cpoe_last",
        "compl": "compl_last",
        "exp_completion": "exp_completion_last",
    },
    inplace=True,
)
# add 1 to season
cpoe_last_py["season"] += 1

# merge together
cpoe_lag_py = cpoe_now_py.merge(
    cpoe_last_py, how="inner", on=["passer_id", "passer", "season"]
)

## look at pass completion stability
cpoe_lag_py[["compl_last", "compl"]].corr()

## look at cpoe stability
cpoe_lag_py[["cpoe_last", "cpoe"]].corr()
